### PR TITLE
check if boxes have logical coordinates (different than a line or a p…

### PR DESCRIPTION
**Pull request by MoussaMM**
This is related to the first bug in this issue https://github.com/tensorflow/tensorflow/issues/35482 .This code fix obliges the user to only pass logical box coordinates. In fact the two diagonal points shouldn't have x1 = x2 or y1 = y2. In this case an invalidArgumentError will be displayed. The complexity of the added code is O(n) which is not influencial in comparaison to the complexity of the non_max_suppression operation : O(n**2).

### CODE REVIEW

**Comment by MoussaMM** [comment]
@googlebot I signed it!

**Comment by rthadur** [comment]
@MoussaMM thank you for your contrbution , i don't see a CLA signed , can you please check again?

**Comment by MoussaMM** [comment]
@rthadur thank you for your reply. I guess i signed with a wrong email. Could you check now please?

**Comment by rthadur** [comment]
@MoussaMM it still says the CLA is not signed , please check again.

**Comment by MoussaMM** [comment]
@rthadur I changed the mail privacy on github so I guess now it works. Can you please check it out again?

File: tensorflow/core/kernels/non_max_suppression_op.cc [hunk 1]
```diff
@@ -39,6 +39,17 @@ namespace {
 
 typedef Eigen::ThreadPoolDevice CPUDevice;
 
+// This function checks that all boxes have coherant ccordinates
+template <typename T>
+bool CheckBoxesCoordinates(const Tensor& boxes, int num_boxes){
+  typename TTypes<T, 2>::ConstTensor boxes_data = boxes.tensor<T, 2>();
+  for(int i=0; i<num_boxes; ++i){
+    if(boxes_data(i, 0) == boxes_data(i, 2) || boxes_data(i, 1) == boxes_data(i, 3))
+      return (false);
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/non_max_suppression_op.cc]: No need to include `false` in parantheses

File: tensorflow/core/kernels/non_max_suppression_op.cc [hunk 2]
```diff
@@ -39,6 +39,17 @@ namespace {
 
 typedef Eigen::ThreadPoolDevice CPUDevice;
 
+// This function checks that all boxes have coherant ccordinates
+template <typename T>
+bool CheckBoxesCoordinates(const Tensor& boxes, int num_boxes){
+  typename TTypes<T, 2>::ConstTensor boxes_data = boxes.tensor<T, 2>();
+  for(int i=0; i<num_boxes; ++i){
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/non_max_suppression_op.cc]: Space between `for` and `(`; space around binary operators

File: tensorflow/core/kernels/non_max_suppression_op.cc [hunk 3]
```diff
@@ -522,6 +533,8 @@ class NonMaxSuppressionOp : public OpKernel {
     int num_boxes = 0;
     ParseAndCheckBoxSizes(context, boxes, &num_boxes);
     CheckScoreSizes(context, num_boxes, scores);
+    // check if the boxes have logical coordinates
+    OP_REQUIRES(context, CheckBoxesCoordinates<float>(boxes, num_boxes), errors::InvalidArgument("boxes coordinates shouldn't have x1=x2 or y1=y2"));
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/non_max_suppression_op.cc]: I would rewrite this to one of:

- `Boxes are empty (x1 == x2 or y1 == y2)`
- `Boxes have a dimension of size 0 (x1 == x2 or y1 == y2)`

Or the same options but without the condition in the paranthesis.

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/non_max_suppression_op.cc (hunk 3 above)
Comment (MoussaMM) [reply tensorflow/core/kernels/non_max_suppression_op.cc]: Thank you for your advice. I corrected those mistakes.

File: tensorflow/core/kernels/non_max_suppression_op.cc [hunk 4]
```diff
@@ -39,6 +39,17 @@ namespace {
 
 typedef Eigen::ThreadPoolDevice CPUDevice;
 
+// This function checks that all boxes have coherant ccordinates
+template <typename T>
+bool CheckBoxesCoordinates(const Tensor& boxes, int num_boxes){
+  typename TTypes<T, 2>::ConstTensor boxes_data = boxes.tensor<T, 2>();
+  for (int i = 0; i < num_boxes; ++i){
+    if(boxes_data(i, 0) == boxes_data(i, 2) || boxes_data(i, 1) == boxes_data(i, 3))
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/non_max_suppression_op.cc]: This line still fails coding style: space after `if`, line length > 80 characters

File: tensorflow/core/kernels/non_max_suppression_op.cc [hunk 5]
```diff
@@ -569,6 +582,8 @@ class NonMaxSuppressionV2Op : public OpKernel {
     int num_boxes = 0;
     ParseAndCheckBoxSizes(context, boxes, &num_boxes);
     CheckScoreSizes(context, num_boxes, scores);
+    // check if the boxes have logical coordinates
+    OP_REQUIRES(context, CheckBoxesCoordinates<T>(boxes, num_boxes), errors::InvalidArgument("boxes coordinates shouldn't have x1=x2 or y1=y2"));
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/non_max_suppression_op.cc]: You should replace all the checks in the same way, not just the one I commented on on the previous review.

Also, the comments above these checks are redundant, please remove.

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
You still haven't changed all `OP_REQUIRES` lines to be consistent. Please, don't only change lines that were commented on.

**Comment by MoussaMM** [comment]
Sorry about my carelessness.

**Review by mihaimaruseac — DISMISSED** [review]

**Comment by MoussaMM** [comment]
@mihaimaruseac multiple tests contain duplicate x or y. What should I do?

**Comment by mihaimaruseac** [comment]
We should fix them: either expect the failure or change the coordinates. Let's do that in this PR too.

**Comment by MoussaMM** [comment]
@mihaimaruseac sorry for bothering but I am quite lost with the testing used by tensorflow. Could you explain me the difference between the tests localized in python folder and the tests localized in compiler/test?

**Comment by mihaimaruseac** [comment]
For the second question, please rebase against master.

For the first question, the compiler/test tests test various compilers to optimize models whereas the python tests are main TF tests.

**Comment by MoussaMM** [comment]
@mihaimaruseac  is there anyway to test if the build works on google cloud instead of my local computer?

**Comment by mihaimaruseac** [comment]
The only way to test on Google infrastructure is after review. But usually we should not have many changes between internal Google and outside (except some tests which I'll take care of during final integration).

Can you solve the conflicts and then ping me again for a new review?

**Review by mihaimaruseac — APPROVED** [review]

**Comment by MoussaMM** [comment]
@mihaimaruseac so how to check the source code of windows failed tests?

**Comment by mihaimaruseac** [comment]
That was an internal failure, let's retrigger

**Comment by MoussaMM** [comment]
It is working! Thank you @mihaimaruseac . It was a pleasure to work with you

**Comment by mihaimaruseac** [comment]
It is now imported internally in Google. There, it gets integrated with internal code and then reviewed again. After that, if all tests pass it will get merged automatically.

Thank you for the PR.

**Comment by mihaimaruseac** [comment]
Hi @MoussaMM The PR has been automatically rolled back in 525e091431b6e6661f5ec6b61b83857791aff441 as it breaks a large number of tests.

We can try a different approach where we add a flag guarding the check for logical coordinates. Or, I will try to fix this internally, though it will take some time and probably won't be in time for the upcoming TF 2.2 release.

**Comment by MoussaMM** [comment]
Hi @mihaimaruseac . Sorry can you explain your approach more? 
Does it mean we are not allowed to change tests and we have to change the behaviour of the algorithm to be coherent with the tests?

**Comment by mihaimaruseac** [comment]
Since this PR is a behavior change, we have two options

1. Either someone internally at Google fixes the behavior everywhere and gets approval from all affected teams. Then we make the fix and we include a note about it in the release notes so that people know to update their scripts in case they will be broken by this.
1. Or, we add a new parameter to `non_max_suppression_padded` which is `False` by default and only checks for logical coordinates if set to `True`. Will also require API review and addition to the release notes.

Both of these might take longer than the time before the next branch cut, however.

**Comment by MoussaMM** [comment]
If you want we can let the illogical boxes get accepted by the algorithm and just eleminate the wrong behaviour they introduce. In fact, even the tests containing illogical boxes didn't have illogical results so I guess it can be merged. 
Nevertheless, I think the error check make more sense. 
(The illogical behaviour I m talking about happens when an illogical boxe get chosen, the algorihm keep rechoosing that box until it reaches the maximum number of result boxes)

**Pull request merged**